### PR TITLE
1st step

### PR DESCRIPTION
**Pull request by JJSwigut**
Started with the first card but I couldn't figure out how to make dummy data work, so I started building the database stuff and then I had to head to work! 
I feel confused by the adapter, viewholder process. I've done it before but every time feels like a struggle. 
Thanks

### CODE REVIEW

**Comment by j-roskopf** [comment]
just a note - it looks like you committed your code to the master branch and you're looking to merge that code into a new branch called starter, usually it will be the other way around. you'll want to create a branch from master, merge your code into there, and then do a PR merging code from that new branch into master

File: app/build.gradle [hunk 1]
```diff
@@ -22,16 +24,43 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
         }
     }
+    buildFeatures {
+        dataBinding true
+    }
 }
 
 dependencies {
+    api 'com.google.android.material:material:1.3.0-alpha02'
```
Comment (j-roskopf) [inline app/build.gradle:33]: how come bringing this in via `api`  instead of `implementation`

File: app/build.gradle [hunk 2]
```diff
@@ -22,16 +24,43 @@ android {
...
     implementation 'androidx.core:core-ktx:1.3.1'
+
+
+    // Navigation
+    implementation "android.arch.navigation:navigation-fragment-ktx:$navigationVersion"
+    implementation "android.arch.navigation:navigation-ui-ktx:$navigationVersion"
+
     implementation 'androidx.appcompat:appcompat:1.2.0'
     implementation 'androidx.constraintlayout:constraintlayout:2.0.1'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
Comment (j-roskopf) [inline app/build.gradle:45]: was legacy supported needed for something?

File: app/src/main/java/com/jjswigut/matters/MainActivity.kt [hunk 3]
```diff
@@ -2,10 +2,18 @@ package com.jjswigut.matters
 
 import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
+import com.jjswigut.matters.ui.MatterListFragment
 
+//ready to go
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/MainActivity.kt:7]: still needed?

File: app/src/main/java/com/jjswigut/matters/database/Matter.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        @PrimaryKey(autoGenerate = true)
+        var matterId:Long = 0L,
+
+        @ColumnInfo(name = "matter_title")
+        val matterTitle:String  ,
+
+        @ColumnInfo(name = "matter_content")
+        var matterContent:String  ,
+
+
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/Matter.kt:20]: just a nit about formatting but these extra lines can be cleaned up

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
+
+
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:2]: extra lines can be cleaned up

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    @Insert
+    suspend fun insert(matter: Matter)
+
+    @Insert
+    suspend fun update(matter: Matter)
+
+    @Query("SELECT * from matter_table WHERE matterId = :key")
+    suspend fun get(key: Long): Matter?
+
+    @Query("DELETE FROM matter_table")
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:22]: I believe room gives us a `Delete` annotation?

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    suspend fun clear()
+
+    @Query("SELECT * FROM matter_table ORDER BY matterId DESC LIMIT 1 ")
+    suspend fun getMatter(): Matter?
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:26]: how is this one going to be used in the future? It would seem to only return 1 note?

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import androidx.room.Insert
+import androidx.room.Query
+
+@Dao
+interface MatterDataBaseDao{
+
+    @Insert
+    suspend fun insert(matter: Matter)
+
+    @Insert
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:16]: I believe there is also an update annotation that room gives us

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+import androidx.lifecycle.LiveData
+import androidx.room.Dao
+import androidx.room.Insert
+import androidx.room.Query
+
+@Dao
+interface MatterDataBaseDao{
+
+    @Insert
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:13]: if we insert a matter / note that already exists, what do we want to happen? might want to specify a conflict resolution strategy

File: app/src/main/java/com/jjswigut/matters/database/MatterDatabase.kt [hunk 10]
```diff
@@ -0,0 +1,38 @@
+
+
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/database/MatterDatabase.kt:2]: extra lines

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+ */
+class MatterListFragment : Fragment() {
+
+    private var columnCount = 1
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.fragment_matter_list, container, false)
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:26]: I know it wasn't in our original plan, but if you wanted to switch over to https://developer.android.com/topic/libraries/view-binding?authuser=2 that would be nice!

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+        // Set the adapter
+        if (view is RecyclerView) {
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:29]: is this ever true? it looks like the root view inside of `fragent_matter_list` is a RelativeLayout so i don't think this would ever be true

File: app/src/main/java/com/jjswigut/matters/ui/MatterListRecyclerViewAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        holder.contentView.text = item.matterContent
+    }
+
+    override fun getItemCount(): Int = values.size
+
+    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val titleView: TextView = view.findViewById(R.id.matter_title)
+        val contentView: TextView = view.findViewById(R.id.content)
+
+        override fun toString(): String {
```
Comment (j-roskopf) [inline app/src/main/java/com/jjswigut/matters/ui/MatterListRecyclerViewAdapter.kt:38]: how come you're overriding toString?

File: app/build.gradle [hunk 14]
```diff
@@ -22,16 +24,43 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
         }
     }
+    buildFeatures {
+        dataBinding true
```
Comment (j-roskopf) [inline app/build.gradle:28]: are you using data binding? I didn't see it used anywhere

File: app/src/main/res/layout/fragment_matter_list.xml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+    <RelativeLayout
```
Comment (j-roskopf) [inline app/src/main/res/layout/fragment_matter_list.xml:2]: this file looks formatted incorrectly.

File: app/build.gradle (hunk 14 above)
Comment (JJSwigut) [reply app/build.gradle:28]: I was going to use it with the viewmodel(that doesn't exist yet) to update the ui's directly and probably for clicklisteners as well

File: app/build.gradle (hunk 1 above)
Comment (JJSwigut) [reply app/build.gradle:33]: I copied it from a google tutorial app. I wasn't sure myself the difference between api and implementation?

File: app/build.gradle (hunk 2 above)
Comment (JJSwigut) [reply app/build.gradle:45]: probably not, I'll get rid of it and find out

File: app/src/main/java/com/jjswigut/matters/MainActivity.kt (hunk 3 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/MainActivity.kt:7]: nope!

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt (hunk 5 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:2]: got it

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 12 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:29]: I'm gonna try this another way. Again, I don't really understand how to do this is in the most simple way because every sample I look at seems to have an entirely different way of setting things up

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 11 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:26]: done!

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt (hunk 7 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:26]: I was thinking about using this to retrieve the note that is clicked on when we navigate to the edit part?_

File: app/build.gradle (hunk 1 above)
Comment (j-roskopf) [reply app/build.gradle:33]: good question! an android app can contain many modules. `app` is just the name of the default module where the application lives. but think of a module like a folder where you can have code live - a larger android app might have 10,20, 100, 500+ modules. a very simple example that I tend to have in most of my app are three modules, an `app` module, a `network` module where my networking code lives, and a `database` module where most of my database operations occur. 

the benefit of having all of these modules, especially in a large code base, is you do not have to recompile the entire code base every time you make a change on one screen. in my 3 module setup for example, if i just make changes to the `network` module, I do not have to recompile the `database` module.

So in this type of multi module setup, when you bring in dependencies into each module, you can usually do it in 2 ways, via `api` or via `implementation`. When you bring in a dependency via `implementation` you are saying i don't want any module that depends on me to see this dependency.  So for example let's say we have module A, B, C, and D

if your dependency graph looked something like

A depends on B, B depends on C, C depends on D 

and you brought each one of those in via implementation, module A knows nothing about module C. And that's good! module A doesn't need to know how module B does its business.

But in that same scenario, if you brought each module into the other via `api`, the above statement would not be true. Module A would know about module C. which can be a good thing if that's the way your project is setup.

But all of that to say in a single module app, we don't need to use `api` for the time being :)

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt (hunk 7 above)
Comment (j-roskopf) [reply app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:26]: I'd imagine we would need to order by ID in our SQL statement though. this is just going to always give us the first note in the DB which I don't think is our desired behavior

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 11 above)
Comment (j-roskopf) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:26]: awesome! ViewBinding is definitely the way to go over findViewbyId / kotlin synthetics.

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 12 above)
Comment (j-roskopf) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:29]: completely understandable! let me know and we can go over how i'd do it as well - but only if ya want to!

File: app/build.gradle (hunk 14 above)
Comment (j-roskopf) [reply app/build.gradle:28]: gotchya - DataBinding is a pretty polarizing thing. No one seems to really be just 'okay' with it - you either hate it / love it.

I'll be curious to hear your opinion on it after a project or two with / without data binding.

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt (hunk 6 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:22]: So I tried to change this but it said that I needed parameters. I looked at the documentation for Room but wasn't really sure what I was looking for?

File: app/src/main/java/com/jjswigut/matters/ui/MatterListRecyclerViewAdapter.kt (hunk 13 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListRecyclerViewAdapter.kt:38]: Another thing I didn't understand and got rid of.

File: app/build.gradle (hunk 1 above)
Comment (JJSwigut) [reply app/build.gradle:33]: Awesome! That makes sense. And when you talk about modules, you're talking higher level than the network package or the database package I may have in the app module?

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 11 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:26]: I understand why it's better than findviewbyid(memory cost searching the whole project everytime) but why is it better than kotlin synthetics?

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 12 above)
Comment (JJSwigut) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:29]: I'd love to know how you would do it! I think I figured out a fairly simple way in my recent PR but  you let me know!

File: app/build.gradle (hunk 14 above)
Comment (JJSwigut) [reply app/build.gradle:28]: When I learned about it I had two thoughts. One, it seemed to clean up the code a lot and reduce the amount of lines that went into things, but two , I think it makes the code less readable because you can't read what's going on the main code. A lot of it is now hidden in the XML and issues could be easier to miss. I think I probably wouldn't use it much, I just thought it was the way things were going

Comment (j-roskopf) [reply app/build.gradle:28]: IMO Databinding is on its way out - the 2 things that really irk me about it is 1 like you mentioned, things hidden in XML. which i think makes it less readable, but also infinitely less testable. any logic that now lives in your XML is 100% not unit testable anymore. my second gripe with it is the build errors are pretty cryptic and the stack traces are pretty hard to parse through

File: app/build.gradle (hunk 1 above)
Comment (j-roskopf) [reply app/build.gradle:33]: right, the network package is just a folder in your app module. you can extract that network folder in your app module to its own network module and have your app module bring in your network module.

Here is a sample base project I created that allows one to have features live in their own module and is a pretty good reference for a small project to have good modularization https://github.com/j-roskopf/BaseAndroidProject

for these types of projects, it's usually overkill to go all out on modularization and doesn't really matter until you get pretty large projects

File: app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt (hunk 6 above)
Comment (j-roskopf) [reply app/src/main/java/com/jjswigut/matters/database/MatterDataBaseDao.kt:22]: ah, I had originally misread. you're looking to delete the entire table and not just an entry from the table - my bad!

File: app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt (hunk 11 above)
Comment (j-roskopf) [reply app/src/main/java/com/jjswigut/matters/ui/MatterListFragment.kt:26]: the main benefit that comes to mind is nullability. you can reference a kotlin synthetic in any fragment / activity (most of the time by accident) and there are no compile time guarantees that you aren't making a mistake. so if you have a textview in fragment A, you can reference that textview in fragment B and you'll get a crash at runtime. viewbinding makes that not possible

File: app/build.gradle (hunk 14 above)
Comment (JJSwigut) [reply app/build.gradle:28]: Word. That makes sense. I'll work without it! I need readable and testable.
